feat: Bandits part 3: Bandit model DTOs

typotter opened the pull request:
### Motivation and context
In order to compute bandit actions, the model data needs to be loaded and parsed. This simple PR introduces the data types needed for Bandit evaluation with built-in parsing from decoded JSON.

### Description
- DTOs for Bandit models and evaluation/log event.
- Parsing built into DTOs
- DTOs for passing user data into the Client (subject and action attributes) as well as convenience methods to make passing data into the client more flexible for the end-user (developers)

### How is this tested
- Unit tests for the helper parsing logic we get in `AttributeSet` and `ContextAttributes`
`BanditDTOTest` parses and verifies a whole tree of data.

typotter commented on src/Bandits/BanditVariationIndexer.php:
```diff
@@ -41,20 +41,19 @@ public function __construct(array $banditVariations)
     public function isBanditFlag($flagKey): bool
     {
-        return array_key_exists($flagKey, $this->banditFlags);
+        return isset($this->banditFlags[$flagKey]);
```
faster resolution than array_key_exists.

typotter commented on src/DTO/Bandit/ActionCoefficients.php:
```diff
@@ -0,0 +1,75 @@
+        public readonly array $actionNumericCoefficients = [],
+        public readonly array $actionCategoricalCoefficients = []
+    ) {
+        // Since PHP doesn't enforce array typing and this dev spent an unfortunate amount of time debugging, we ensure
```
higher levels of DTOs properly enforce the typings

aarsilv replied:
I love how the comment alludes to this decision being made from blood sweat and tears 😅

typotter commented on src/DTO/Bandit/BanditEvaluationEvent.php:
```diff
@@ -0,0 +1,47 @@
+use DateTime;
+use JsonSerializable;
+
+class BanditEvaluationEvent implements JsonSerializable
```
many of the telemetry packages accept a dictionary/map-like structure for userdata.

aarsilv commented on src/DTO/Bandit/AttributeSet.php:
```diff
@@ -0,0 +1,56 @@
+    ) {
+    }
+
+    public static function fromArray(array $attributes): self
```
❤️

aarsilv commented on src/DTO/Bandit/AttributeSet.php:
```diff
@@ -0,0 +1,56 @@
+
+    /**
+     * Types to automatically classify as Numeric
+     * `float` and `double` are the sane datatype in php.
+     * In PHP, for historical reasons, `gettype` returns `double`.
```
I appreciate the comment--I was just wondering if the two were different somewhow!

aarsilv commented on src/DTO/Bandit/IContextAttributes.php:
```diff
@@ -0,0 +1,29 @@
+{
+    public function getAttributes(): AttributeSet;
+
+    public function getKey(): string;
```
What do we need `$key` for? Each attribute should have it's own key.

typotter replied:
`ContextAttributes` pairs a string `key` with an `AttributeSet` in order to keep SubjectIDs and ActionKeys closer to their attributes when dealing with groups of same. It's a convenience layer that alleviates some awkwardness with having to access both key and value when manipulating arrays.

I've updated the docs on what exactly the `IContextAttributes` interface is intended for/what problem it solves. Hopefully that makes it clear why the added layer on top of `AttributeSet`

aarsilv commented on tests/DTO/Bandit/AttributeSetTest.php:
```diff
@@ -0,0 +1,70 @@
+use Eppo\DTO\Bandit\AttributeSet;
+use PHPUnit\Framework\TestCase;
+
+class AttributeSetTest extends TestCase
```
🙌

aarsilv commented on tests/DTO/Bandit/AttributeSetTest.php:
```diff
@@ -0,0 +1,70 @@
+        $this->assertEquals($expectedCategoricalAttributes, $attributeSet->categoricalAttributes);
+    }
+
+    public function testFromArrayFiltersInvalidValueTypes()
```
🔥

aarsilv commented on tests/DTO/BanditDTOTest.php:
```diff
@@ -0,0 +1,164 @@
+
+        $this->assertNotNull($bandit);
+
+        // Assert on most properties down the entire object tree, testing the fromJSON methods of each DTO.
```
💪

aarsilv reviewed (approved):
Looking good! My only substantive comment is that I don't think `ContextAttributes` needs its own key given all of the attributes nested within have their keys.

typotter commented:
Thanks @aarsilv . I addressed your comments on `ContextAttributes::$key` and added filtering in `AttributeSet` to pre-emptively filter out non-numerics.
